Fixes to iOS render loop and HUD overlay.

billhollings opened the pull request:
* Refactor Platform::main_loop_frame() and main_loop(),
  to reduce fragility, and avoid unnecessary code duplication.

* Avoid initial app_requested() test in main_loop_frame(),
  as it was causing looping to stop after one frame.

* Since main_loop() is not a virtual function, remove
  declaration comment indicating it can be overridden.

* Set iOS view to have a black background, to ensure 
  transparency in HUD overlay displays in full on device.
  
* Fix rendering scaling in IosWindow::get_dpi_factor().
  Previous value was causing HUD overlay to render magnitudes too large.
  
* Add macOS files to .gitignore.

I created this from a branch on this repo instead of a separate fork, because I was not expecting to need to make a fix.

Fixes #1121
Fixes #1131

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Copyright-Notice-and-License-Template)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes do not add any regressions
- [ ] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

 Note: The Samples CI runs a number of checks including:
 - [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
 - [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)

 If this PR contains framework changes:
 - [ ] I did a full batch run using the `batch` command line argument to make sure all samples still work properly

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [ ] I have tested the sample on at least one compliant Vulkan implementation
- [ ] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)
- [ ] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [ ] Any dependent assets have been merged and published in downstream modules
- [ ] For new samples, I have added a paragraph with a summary to the appropriate chapter in the readme of the folder that the sample belongs to [e.g. api samples readme](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/samples/api/README.adoc)
- [ ] For new samples, I have added a tutorial README.md file to guide users through what they need to know to implement code using this feature. For example, see [conditional_rendering](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/samples/extensions/conditional_rendering)
- [ ] For new samples, I have added a link to the [Antora navigation](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/antora/modules/ROOT/nav.adoc) so that the sample will be listed at the Vulkan documentation site

asuessenbach commented on framework/platform/platform.h:
```diff
@@ -130,6 +128,8 @@ class Platform
 	static const uint32_t MIN_WINDOW_HEIGHT;
 
   protected:
+	virtual bool uses_internal_looping() { return true; }
```
Maybe make this function `const`?

billhollings replied:
Fixed.

asuessenbach commented on framework/platform/ios/ios_platform.h:
```diff
@@ -33,6 +33,7 @@ class IosPlatform : public Platform
     void* view;
 
 protected:
+	bool uses_internal_looping() override { return false; }
```
What does this `internal_looping` mean? Why is it different on IOS than everywhere else?

billhollings replied:
On other platforms, `main_loop()` is called once. On iOS, `main_loop()` is called on each frame by the platform display synchronization loop. See `ViewController::renderLoop`. This is generally how rendering is done on iOS.

This was apparently why `Platform::main_loop_frame()` was added separately from `Platform::main_loop()`, but this code replication is now showing it's maintainability problems in that maintaining one loop code and copying into the other loop code will sometimes cause unexpected behaviour.

Code had recently been added to the top of `main_loop_frame()` that assumed it would only be called once, which then broke on iOS.

Another option would be to separate the updating/rendering code inside the loop, from `main_loop()`, so it can be called separately on iOS. I decided a loop check would be a more minimal change, but I can easily take that approach, since it's probably a better factoring.

asuessenbach replied:
ok... what do you think about having the one-time code in `main_loop()` and then calling into `main_loop_frame()` in a loop? Would that work?

billhollings replied:
Good idea! I've implemented this now.

asuessenbach commented:
You should at least resolve the failing Quality Check.

SRSaunders committed:
Fixes to iOS render loop and HUD overlay.

* Refactor Platform::main_loop_frame() and main_loop(),
  to reduce fragility, and avoid unnecessary code duplication.

* Avoid initial app_requested() test in main_loop_frame(),
  as it was causing looping to stop after one frame.

* Since main_loop() is not a virtual function, remove
  declaration comment indicating it can be overridden.

* Set iOS view to have a black background, to ensure
  transparency in HUD overlay displays in full on device.

* Fix rendering scaling in IosWindow::get_dpi_factor().
  Previous value was causing HUD overlay to render magnitudes too large.

* Add macOS files to .gitignore.

billhollings commented:
> You should at least resolve the failing Quality Check.

If tried many times.

1. For some reason, `clang-format` works slightly differently on macOS than on other platforms, so running `./scripts/clang_format.py main` ends up changing more than expected, and breaks the CI test here.
2. The [CI error messages](https://github.com/KhronosGroup/Vulkan-Samples/actions/runs/10461719961/job/28970635867?pr=1120) are incomprehensible to me, so I'm not able to decipher what it's complaining about, in order to simply fix it manually.

gpx1000 commented:
https://github.com/KhronosGroup/Vulkan-Samples/actions/runs/10461719961?pr=1120

If you scroll to the bottom you can grab the diff that can be directly run and hopefully fix the issue.

asuessenbach commented:
> For some reason, clang-format works slightly differently on macOS than on other platforms

Indeed, it should remove white spaces at the end of a line, and it should replace spaces with tabs.
This patch shows the difference I would get when running clang-format: [0001-formatting.patch](https://github.com/user-attachments/files/16670591/0001-formatting.patch)

billhollings commented:
@gpx1000 @asuessenbach 

Thanks for the help on code formatting. I've got this formatting correctly now.

Plus I've added a couple of other iOS fixes, and updated the title and description above.

gpx1000 commented on framework/platform/ios/ios_window.mm:
```diff
@@ -69,7 +69,7 @@
 float IosWindow::get_dpi_factor() const
 {
-    return [[UIScreen mainScreen] nativeScale] * (([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPad) ? 132 : 163);
+	return 1.0;
```
This is almost nothing more than a question; Is this considered the "right way" to get the dpi for text for all iOS devices?  It seems that a 1-1 ratio might be wrong in some cases?  I guess not given the displays themselves are small and the pixel pitch is small.

SRSaunders replied:
I think @billhollings and I found the same issue that was causing UI Overlay scaling problems on iOS.  The way I solved it was the following.  This works for real iOS devices and the iOS Simulator:

In ViewController.mm:

```
-->	self.vulkan_view.contentScaleFactor = UIScreen.mainScreen.nativeScale;
        context = create_platform_context((int)args.count, const_cast<char**>(argv));
        _context = (vkb::IosPlatformContext*)context.get();
        _context->view = self.vulkan_view;
        _code = (vkb::ExitCode)platform_main(*context);
        uint32_t fps = 60;
```

In ios_window.mm

`return [[UIScreen mainScreen] nativeScale];`

SRSaunders replied:
Looking at this again, I think the following is a bit more semantically correct:

In **ViewController.mm**:

```
...
-->	self.vulkan_view.contentScaleFactor = UIScreen.mainScreen.nativeScale;
        context = create_platform_context((int)args.count, const_cast<char**>(argv));
        _context = (vkb::IosPlatformContext*)context.get();
...
```

In **ios_window.mm**:

```
float IosWindow::get_dpi_factor() const
{
	return 1.0;
}

float IosWindow::get_content_scale_factor() const
{
	return [[UIScreen mainScreen] nativeScale];
}
```

I think @billhollings is correct to return 1.0 for `get_dpi_factor()`.  My changes for setting and getting the view's `contentScaleFactor` are also okay, but they offset each other for the imgui UI Overlay and the code would still work without adding this part.

billhollings replied:
> I think @billhollings is correct to return 1.0 for `get_dpi_factor()`. My changes for setting and getting the view's `contentScaleFactor` are also okay, but they offset each other for the imgui UI Overlay and the code would still work without adding this part.

Okay. How about I leave it as it is now (returning `1.0`), and adding additional logic around screen attributes can be left for another PR?

It is currently displaying nicely on both iPhone and iPad. On an iPad, the app display Idiom seems to stay as iPhone, so it uses the standard iPad PiP when an app is only designed for iPhone. As a result, it looks the same on both iPhone and iPad.

SRSaunders replied:
> How about I leave it as it is now (returning 1.0), and adding additional logic around screen attributes can be left for another PR?

That's fine for now, but the iOS view `contentScaleFactor` and `get_content_scale_factor()` need to be setup for other samples in the project (e.g. _swapchain_recreation_).  I can submit these changes after your PR is merged.

I will work on your review request tomorrow.  There's an additional error path that needs to be handled in `main_loop_frame()` when the sample is not properly specified (e.g. cmd line typo). Should be a simple fix.

asuessenbach commented on framework/platform/platform.cpp:
```diff
@@ -185,49 +177,7 @@ ExitCode Platform::main_loop()
-				return ExitCode::FatalError;
-			}
-		}
+		main_loop_frame();
```
You're missing any failure exit code from main_loop_frame.
Maybe something like
```
ExitCode Platform::main_loop()
{
	ExitCode exit_code = ExitCode::Success;
	if (!app_requested())
	{
		exit_code = ExitCode::NoSample;
	}

	while ((exit_code == ExitCode::Success) && !window->should_close() && !close_requested)
	{
		exit_code = main_loop_frame();
	}

	return exit_code;
}

```

billhollings replied:
D'oh! That's my bad. Thanks for catching it. Fixed now.

I've left the initial `!app_requested()` test as an immediate return, to clarify it's use as a one-time pre-check, and keep it disentangled from the looping logic.

SRSaunders commented on framework/platform/platform.cpp:
```diff
@@ -121,55 +121,47 @@ ExitCode Platform::initialize(const std::vector<Plugin *> &plugins)
-				on_app_close(id);
-				active_app->finish();
-			}
+		update();
```
To handle the case when `app_requested()` returns false right at the start (e.g. due to a typo in the sample name), you should probably do the following:

```
...
			active_app->update(0.01667f);
		}

-->		if (!active_app)
-->		{
-->			return ExitCode::NoSample;
-->		}

		update();

-->		if (active_app->should_close())
		{
			std::string id = active_app->get_name();
...
```

This will prevent things from crashing on an undefined `active_app` inside `update()`.  Note this is not an issue with non-iOS builds since an active sample is already validated in `platform_main()`.  However, for iOS the return code is not checked in `ViewController::viewDidLoad` and that is the root cause.  Nonetheless, the `main_loop_frame()` code should be tolerant of this, and at least not crash inside `update()` with an undefined `active_app`.  The iOS `viewDidLoad` code could be fixed later, along with a way to actually terminate the iOS app if the sample never loaded (or a batch run finishes).

SRSaunders commented on .gitignore:
```diff
@@ -16,3 +16,6 @@ tags
 .tracy
+
+# macOS
+.DS_Store
```
.DS_Store now appears twice in the .gitignore file.  You could either remove the earlier instance, or revert your change - either approach is fine by me.

asuessenbach reviewed (approved):
By code inspecting and running on Windows: LGTM.

SRSaunders reviewed (approved):
LGTM.  Did full batch runs for iOS + macOS and both `main_loop_frame()` and `main_loop()` worked correctly.  Still need fixes for non-success return codes in iOS `viewDidLoad()` and `renderLoop()`, but these can be dealt with in a future PR.

billhollings commented:
> LGTM. Did full batch runs for iOS + macOS and both `main_loop_frame()` and `main_loop()` worked correctly. Still need fixes for non-success return codes in iOS `viewDidLoad()` and `renderLoop()`, but these can be dealt with in a future PR.

@SRSaunders Would you mind giving this a formal approval, so it can be merged, please?

Once the approvals are in, what is the process for merging? Do I do that? Or is there a designated person responsible for merging?

gpx1000 commented:
We usually ask @marty-johnson59 to handle the merging.  We usually merge on 3; but Marty, I think this is safe to merge on 2.

SRSaunders commented:
> @SRSaunders Would you mind giving this a formal approval, so it can be merged, please?

@billhollings I don't have approval rights for this project.  I am happy to provide feedback, but I cannot mark the PR as ready for merge.  As you see above, @gpx1000 has that ability.

marty-johnson59 commented:
Thanks.  Given this is Mac (we have limited folks who can test), I'm happy to merge w/ 2.  We can revert if needed.
